### PR TITLE
Don't unregister the TimerSet and timers if exception is thrown in Ti…

### PR DESCRIPTION
**Pull request by nemanja-boric-sociomantic**
…merExt

`TimerExt` is now swallowing all the unhandled exceptions, keeping the timers
registered by default. Previously, the uncaught exception would
propagate to TimerSet, unregistering it and making all TimerExt
registration not work again. That behaviour is now fixed, with the
default that, if exception is thrown, the timer will stay registered.

### CODE REVIEW

**Comment by nemanja-boric-sociomantic** [comment]
Marked as a breaking change, because users may use `TimerExt` for purposes not covered by `DaemonApp`, so that could change the application's behavior.

**Review by gavin-norman-sociomantic — APPROVED** [review]

**Comment by gavin-norman-sociomantic** [comment]
LGTM, but will leave a bit in case anyone else has any comments.

**Comment by mathias-lang-sociomantic** [comment]
Test ?

File: src/ocean/util/app/ext/TimerExt.d [hunk 1]
```diff
@@ -27,6 +27,19 @@
...
 
+import ocean.util.log.Log;
+
+/*******************************************************************************
+
+    Static module logger
+
+*******************************************************************************/
+
+static private Logger log;
```
Comment (mathias-lang-sociomantic) [inline src/ocean/util/app/ext/TimerExt.d]: Redundant static. Also last time a module-level logger was added it didn't go so well. Could you make it class-local?

File: src/ocean/util/app/ext/TimerExt.d [hunk 2]
```diff
@@ -27,6 +27,19 @@
 
 module ocean.util.app.ext.TimerExt;
 
+import ocean.util.log.Log;
```
Comment (mathias-lang-sociomantic) [inline src/ocean/util/app/ext/TimerExt.d]: `Logger`, `Log` will be gone in v4 and is deprecated in v3 but v4.x.x is too much behind

File: src/ocean/util/app/ext/TimerExt.d (hunk 1 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/ext/TimerExt.d]: bah, I've copy/pasted it from somewhere else.

**Comment by mathias-lang-sociomantic** [comment]
I suggest merging v3 into v4 (fast-forwarding?) before merging anything into it.

**Comment by nemanja-boric-sociomantic** [comment]
> I suggest merging v3 into v4 (fast-forwarding?) before merging anything into it.

@mihails-strasuns-sociomantic said he's going to do this.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with the fix for the unittests, test for the catching behavior and `Log` -> `Logger` import.

File: src/ocean/util/app/ext/TimerExt.d [hunk 3]
```diff
@@ -87,7 +87,7 @@ unittest
         }
     }
 
-    new App;
+    (new App).run(null);
```
Comment (mathias-lang-sociomantic) [inline src/ocean/util/app/ext/TimerExt.d]: Not an unittest please, an integration test. `unittest` shouldn't do IO or affect global state, and `epoll` does.

Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/ext/TimerExt.d]: Well, there's no IO with this particular unittest. The `epoll_create` and `timerfd_create` calls already affect the global state, with one `epoll_wait` cycle or not.

Comment (mathias-lang-sociomantic) [reply src/ocean/util/app/ext/TimerExt.d]: Well, I think they shouldn't be called then :)

Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/ext/TimerExt.d]: Ok, I'll move the test into an integration test and remove `new App` from here.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with removing instantantion of the applications inside unittests and moving the `TimerExt` test into `test/timerext/main.d`

**Comment by nemanja-boric-sociomantic** [comment]
Replaced `e.toString()` with `getMsg(e)`, since the timers may be called multiple times per second, unlike in #227.

File: test/timerext/main.d [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        }
+    }
+
+    private bool third ( )
+    {
+        return false;
+    }
+}
+
+void main(istring[] args)
```
Comment (mathias-lang-sociomantic) [inline test/timerext/main.d]: Missing space

File: test/timerext/main.d [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+{
+    auto sandbox = DirectorySandbox.create(["etc", "log"]);
+
+    File.set("etc/config.ini", "[LOG.Root]\n" ~
+               "console = false\n");
+
+    auto app = new App;
+    app.main(args);
```
Comment (mathias-lang-sociomantic) [inline test/timerext/main.d]: Shouldn't you check the return code also ?

File: test/timerext/main.d (hunk 4 above)
Comment (nemanja-boric-sociomantic) [reply test/timerext/main.d]: :grumpy-cat:

**Comment by nemanja-boric-sociomantic** [comment]
updated: check the code of the `app.main` in test.

**Comment by nemanja-boric-sociomantic** [comment]
I forgot the space >_>.

**Comment by mihails-strasuns-sociomantic** [comment]
Please rebase

**Comment by nemanja-boric-sociomantic** [comment]
Rebased.

**Review by mihails-strasuns-sociomantic — APPROVED** [review]

**Comment by mihails-strasuns-sociomantic** [comment]
Checked coverage stats manually and it seems correct - don't think we can do any better until reports from int tests are sent.

**Pull request merged**